app/models/concerns/profile_fields.rb

```diff
@@ -38,6 +38,17 @@ def canonical?
       parent_profile_id.blank?
     end
 
+    def rules_and_rule_groups
+      arranged_rule_groups = rule_groups.includes(:rules).arrange_serializable do |parent, children|
+        rule_children = parent.rules.map do |r|
+          { rule: r.id }
+        end
+        { rule_group: parent.id, group_children: children, rule_children: rule_children }
+      end
+
+      arranged_rule_groups.concat(rules.without_rule_group_parent.map { |pr| { rule: pr.id } })
+    end
+
```

Comment:
@skateman @romanblanco After working on the fix for the `profile_rule_groups` table and understanding the relationship between the profile, rules and groups more I was able to simplify the ruby solution a good amount. I also moved it to profile_fields.rb so that we can use it for the endpoint. 

This takes about .062s so it's pretty fast and I was able to remove a lot of the maps. I checked the logs and this only runs 5 queries total so a query isn't being ran inside the loop anymore. The unfinished recursive PostgreSQL solution is already around 0.262s and that's only setting up the rule groups in a hierarchical format, not factoring rules in yet or converting to json.

Let me know how this looks to you guys!